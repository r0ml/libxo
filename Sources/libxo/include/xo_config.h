/* libxo/xo_config.h.  Generated from xo_config.h.in by configure.  */
/* libxo/xo_config.h.in.  Generated from configure.ac by autoheader.  */

/* Define to 1 if using 'alloca.c'. */
/* #undef C_ALLOCA */

/* Define to 1 if you have 'alloca', as a function or macro. */
#define HAVE_ALLOCA 1

/* Define to 1 if <alloca.h> works. */
#define HAVE_ALLOCA_H 1

/* Define to 1 if you have the 'asprintf' function. */
#define HAVE_ASPRINTF 1

/* Define to 1 if you have the 'bzero' function. */
#define HAVE_BZERO 1

/* Define to 1 if you have the 'ctime' function. */
#define HAVE_CTIME 1

/* Define to 1 if you have the <ctype.h> header file. */
#define HAVE_CTYPE_H 1

/* Define to 1 if you have the declaration of '__isthreaded', and to 0 if you
   don't. */
#define HAVE_DECL___ISTHREADED 0

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define to 1 if you have the 'dlfunc' function. */
/* #undef HAVE_DLFUNC */

/* Define to 1 if you have the <errno.h> header file. */
#define HAVE_ERRNO_H 1

/* Style of etext */
#define HAVE_ETEXT 2

/* Define to 1 if you have the 'fdopen' function. */
#define HAVE_FDOPEN 1

/* Define to 1 if you have the 'flock' function. */
#define HAVE_FLOCK 1

/* Using real gcc */
/* #undef HAVE_GCC */

/* Define to 1 if you have the 'getpass' function. */
#define HAVE_GETPASS 1

/* Define to 1 if you have the 'getprogname' function. */
#define HAVE_GETPROGNAME 1

/* Define to 1 if you have the 'getrusage' function. */
#define HAVE_GETRUSAGE 1

/* gettext(3) */
/* #undef HAVE_GETTEXT */

/* Define to 1 if you have the 'gettimeofday' function. */
#define HAVE_GETTIMEOFDAY 1

/* humanize_number(3) */
/* #undef HAVE_HUMANIZE_NUMBER */

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <langinfo.h> header file. */
#define HAVE_LANGINFO_H 1

/* Define to 1 if you have the 'crypto' library (-lcrypto). */
/* #undef HAVE_LIBCRYPTO */

/* Define to 1 if you have the 'm' library (-lm). */
#define HAVE_LIBM 1

/* Define to 1 if you have the <libutil.h> header file. */
/* #undef HAVE_LIBUTIL_H */

/* Define to 1 if your system has a GNU libc compatible 'malloc' function, and
   to 0 otherwise. */
#define HAVE_MALLOC 1

/* Define to 1 if you have the 'memmove' function. */
#define HAVE_MEMMOVE 1

/* Define to 1 if you have the 'memrchr' function. */
/* #undef HAVE_MEMRCHR */

/* Define to 1 if you have the <monitor.h> header file. */
#define HAVE_MONITOR_H 1

/* Support printflike */
/* #undef HAVE_PRINTFLIKE */

/* Define to 1 if your system has a GNU libc compatible 'realloc' function,
   and to 0 otherwise. */
#define HAVE_REALLOC 1

/* Define to 1 if you have the 'srand' function. */
#define HAVE_SRAND 1

/* Define to 1 if you have the 'sranddev' function. */
#define HAVE_SRANDDEV 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdio_ext.h> header file. */
/* #undef HAVE_STDIO_EXT_H */

/* Define to 1 if you have the <stdio.h> header file. */
#define HAVE_STDIO_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <stdtime/tzfile.h> header file. */
/* #undef HAVE_STDTIME_TZFILE_H */

/* Define to 1 if you have the 'strchr' function. */
#define HAVE_STRCHR 1

/* Define to 1 if you have the 'strchrnul' function. */
/* #undef HAVE_STRCHRNUL */

/* Define to 1 if you have the 'strcspn' function. */
#define HAVE_STRCSPN 1

/* Define to 1 if you have the 'strerror' function. */
#define HAVE_STRERROR 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the 'strlcpy' function. */
#define HAVE_STRLCPY 1

/* Define to 1 if you have the 'strspn' function. */
#define HAVE_STRSPN 1

/* Have struct sockaddr_un.sun_len */
#define HAVE_SUN_LEN 1

/* Define to 1 if you have the 'sysctlbyname' function. */
#define HAVE_SYSCTLBYNAME 1

/* Define to 1 if you have the <sys/param.h> header file. */
#define HAVE_SYS_PARAM_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/sysctl.h> header file. */
#define HAVE_SYS_SYSCTL_H 1

/* Define to 1 if you have the <sys/time.h> header file. */
#define HAVE_SYS_TIME_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <threads.h> header file. */
/* #undef HAVE_THREADS_H */

/* thread-local setting */
#define HAVE_THREAD_LOCAL THREAD_LOCAL_before

/* Define to 1 if you have the <tzfile.h> header file. */
#define HAVE_TZFILE_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the '__flbf' function. */
/* #undef HAVE___FLBF */

/* Enable debugging */
/* #undef LIBXO_DEBUG */

/* Enable xpath-like filters */
/* #undef LIBXO_NEED_FILTERS */

/* Enable text-only rendering */
/* #undef LIBXO_TEXT_ONLY */

/* Version number as dotted value */
#define LIBXO_VERSION "1.7.5"

/* Version number extra information */
#define LIBXO_VERSION_EXTRA "-git-main"

/* Version number as a number */
#define LIBXO_VERSION_NUMBER 1007005

/* Version number as string */
#define LIBXO_VERSION_STRING "1007005"

/* Enable local wcwidth implementation */
#define LIBXO_WCWIDTH 1

/* Define to the sub-directory where libtool stores uninstalled libraries. */
#define LT_OBJDIR ".libs/"

/* Name of package */
#define PACKAGE "libxo"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "phil@juniper.net"

/* Define to the full name of this package. */
#define PACKAGE_NAME "libxo"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "libxo 1.7.5"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "libxo"

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION "1.7.5"

/* If using the C implementation of alloca, define if you know the
   direction of stack growth for your system; otherwise it will be
   automatically deduced at runtime.
	STACK_DIRECTION > 0 => grows toward higher addresses
	STACK_DIRECTION < 0 => grows toward lower addresses
	STACK_DIRECTION = 0 => direction of growth unknown */
/* #undef STACK_DIRECTION */

/* Define to 1 if all of the C89 standard headers exist (not just the ones
   required in a freestanding environment). This macro is provided for
   backward compatibility; new code need not use it. */
#define STDC_HEADERS 1

/* Use int return codes */
/* #undef USE_INT_RETURN_CODES */

/* Version number of package */
#define VERSION "1.7.5"

/* Retain hash bucket size */
/* #undef XO_RETAIN_SIZE */

/* Define to '__inline__' or '__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif

/* Define to rpl_malloc if the replacement function should be used. */
/* #undef malloc */

/* Define to rpl_realloc if the replacement function should be used. */
/* #undef realloc */

/* Define as 'unsigned int' if <stddef.h> doesn't define. */
/* #undef size_t */
