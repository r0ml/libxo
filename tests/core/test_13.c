/*
 * Copyright (c) 2022, Juniper Networks, Inc.
 * All rights reserved.
 * This SOFTWARE is licensed under the LICENSE provided in the
 * ../Copyright file. By downloading, installing, copying, or otherwise
 * using the SOFTWARE, you agree to be bound by the terms of that
 * LICENSE.
 * Phil Shafer, Feb 2022
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include <langinfo.h>
#include <unistd.h>
#include <wchar.h>
#include <fcntl.h>

#include "xo_config.h"
#include "xo.h"
#include "xo_encoder.h"

static size_t padding_for_month[12];
static size_t month_max_size = 0;

static const char *
get_abmon(int mon)
{

	switch (mon) {
	case 0: return (nl_langinfo(ABMON_1));
	case 1: return (nl_langinfo(ABMON_2));
	case 2: return (nl_langinfo(ABMON_3));
	case 3: return (nl_langinfo(ABMON_4));
	case 4: return (nl_langinfo(ABMON_5));
	case 5: return (nl_langinfo(ABMON_6));
	case 6: return (nl_langinfo(ABMON_7));
	case 7: return (nl_langinfo(ABMON_8));
	case 8: return (nl_langinfo(ABMON_9));
	case 9: return (nl_langinfo(ABMON_10));
	case 10: return (nl_langinfo(ABMON_11));
	case 11: return (nl_langinfo(ABMON_12));
	}

	/* should never happen */
	abort();
}

static void
compute_abbreviated_month_size(void)
{
	int i;
	size_t width;
	size_t months_width[12];

	for (i = 0; i < 12; i++) {
		width = strlen(get_abmon(i));
		if (width == (size_t)-1) {
			month_max_size = -1;
			return;
		}
		months_width[i] = width;
		if (width > month_max_size)
			month_max_size = width;
	}

	for (i = 0; i < 12; i++)
		padding_for_month[i] = month_max_size - months_width[i];
}

static void
printsize(const char *field, size_t width, off_t bytes)
{
	char fmt[BUFSIZ];
	
		/* This format assignment needed to work round gcc bug. */
		snprintf(fmt, sizeof(fmt), "{:%s/%%%dj%sd} ",
		     field, (int) width, "");
		xo_emit_f(XOEF_NO_RETAIN, fmt, (intmax_t) bytes);
}

static size_t
ls_strftime(char *str, size_t len, const char *fmt, const struct tm *tm)
{
	char *posb, nfmt[BUFSIZ];
	const char *format = fmt;
	size_t ret;

	if ((posb = strstr(fmt, "%b")) != NULL) {
		if (month_max_size > 0) {
			snprintf(nfmt, sizeof(nfmt),  "%.*s%s%*s%s",
			    (int)(posb - fmt), fmt,
			    get_abmon(tm->tm_mon),
			    (int)padding_for_month[tm->tm_mon],
			    "",
			    posb + 2);
			format = nfmt;
		}
	}
	ret = strftime(str, len, format, tm);
	return (ret);
}

static void
printtime(const char *field, time_t ftime)
{
	char longstring[80];
	char fmt[BUFSIZ];
	static time_t now = 0;
	const char *format;
	static int d_first = -1;

	if (d_first < 0)
	    d_first = 1;
	if (now == 0)
	    now = time(NULL);

#define	SIXMONTHS	((365 / 2) * 86400)
	if (1)
		/* mmm dd hh:mm || dd mmm hh:mm */
		format = d_first ? "%e %b %R" : "%b %e %R";
	else
		/* mmm dd  yyyy || dd mmm  yyyy */
		format = d_first ? "%e %b  %Y" : "%b %e  %Y";
	ls_strftime(longstring, sizeof(longstring), format, gmtime(&ftime));

	snprintf(fmt, sizeof(fmt), "{d:%s/%%hs} ", field);
	xo_attr("value", "%ld", (long) ftime);
	xo_emit_f(XOEF_NO_RETAIN, fmt, longstring);
	snprintf(fmt, sizeof(fmt), "{en:%s/%%ld}", field);
	xo_emit_f(XOEF_NO_RETAIN, fmt, (long) ftime);
}


int
main (int argc, char **argv)
{
    int i, count = 10;
    xo_emit_flags_t flags = XOF_RETAIN_ALL;
    int opt_color = 1;

    xo_set_program("test_13");

    argc = xo_parse_args(argc, argv);
    if (argc < 0)
	return 1;

    compute_abbreviated_month_size();

    for (argc = 1; argv[argc]; argc++) {
	if (xo_streq(argv[argc], "xml"))
	    xo_set_style(NULL, XO_STYLE_XML);
	else if (xo_streq(argv[argc], "json"))
	    xo_set_style(NULL, XO_STYLE_JSON);
	else if (xo_streq(argv[argc], "text"))
	    xo_set_style(NULL, XO_STYLE_TEXT);
	else if (xo_streq(argv[argc], "html"))
	    xo_set_style(NULL, XO_STYLE_HTML);
	else if (xo_streq(argv[argc], "no-color"))
	    opt_color = 0;
	else if (xo_streq(argv[argc], "pretty"))
	    xo_set_flags(NULL, XOF_PRETTY);
	else if (xo_streq(argv[argc], "xpath"))
	    xo_set_flags(NULL, XOF_XPATH);
	else if (xo_streq(argv[argc], "info"))
	    xo_set_flags(NULL, XOF_INFO);
	else if (xo_streq(argv[argc], "no-retain"))
	    flags &= ~XOF_RETAIN_ALL;
	else if (xo_streq(argv[argc], "big")) {
	    if (argv[argc + 1]) {
		const char *cp = argv[++argc];
		char *ep;
		count = strtoul(cp, &ep, 0);
		if (ep && *ep) {
		    const char suff[] = "kmgt";
		    unsigned long mult[]
			= { 1000, 1000000, 1000000000, 1000000000000 };
		    char *sp = strchr(suff, *ep);
		    if (sp) {
			count *= mult[sp - suff];
		    }
		}
	    }
	} else if (xo_streq(argv[argc], "null")) {
	    int fd = open("/dev/null", O_WRONLY);
	    if (fd >= 0) {
		close(1);
		dup2(fd, 1);
	    }
	}
    }

    xo_set_flags(NULL, XOF_UNITS); /* Always test w/ this */
    if (opt_color)
	xo_set_flags(NULL, XOF_COLOR); /* Force color output */
    xo_set_file(stdout);

    xo_open_container("top");
    xo_open_container("data");

    if (flags != 0)
	xo_set_flags(NULL, flags);

    xo_open_list("entry");

    for (i = 0; i < count; i++) {
	xo_open_instance("entry");

	char name[80];
	snprintf(name, sizeof(name), "xx-%08u", i);

	xo_emitr("{ke:name/%hs}", name);

	xo_emitr("{t:inode/%*ju} ", 3, 12);
	xo_emitr("{t:blocks/%*jd} ", 4, 1234);

	xo_emitr("{t:mode/%s}{e:mode_octal/%03o} {t:links/%*ju} {t:user/%-*s}  {t:group/%-*s}  ",
		"mode", 0660, 2, (uintmax_t) 12,
		4, "phil", 4, "phil");


	xo_emitr("{e:type/%s}", "regular");

	xo_emitr("{:flags/%-*s} ", 3, "123");
	xo_emitr("{t:label/%-*s} ", 4, "1234");
	printsize("size", 5, 12345);
	printtime("modify-time", 1644355825);

	xo_emitr("{dk:name/%hs}", name);

	xo_close_instance("entry");
	xo_emit("\n");
    }

    xo_close_list("entry");

    xo_emit("{Lwc:hits}{:hits/%ld}\n", xo_retain_get_hits());

    xo_close_container("data");
    xo_close_container_h(NULL, "top");

    xo_finish();

    return 0;
}
