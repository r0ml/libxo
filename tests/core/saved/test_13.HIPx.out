<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000000']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000000']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000000']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000000']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000000']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000000']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000000']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000000']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000000']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000000']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000000</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000001']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000001']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000001']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000001']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000001']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000001']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000001']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000001']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000001']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000001']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000001</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000002']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000002']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000002']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000002']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000002']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000002']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000002']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000002']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000002']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000002']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000002</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000003']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000003']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000003']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000003']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000003']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000003']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000003']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000003']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000003']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000003']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000003</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000004']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000004']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000004']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000004']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000004']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000004']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000004']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000004']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000004']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000004']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000004</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000005']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000005']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000005']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000005']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000005']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000005']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000005']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000005']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000005']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000005']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000005</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000006']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000006']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000006']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000006']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000006']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000006']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000006']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000006']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000006']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000006']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000006</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000007']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000007']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000007']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000007']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000007']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000007']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000007']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000007']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000007']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000007']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000007</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000008']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000008']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000008']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000008']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000008']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000008']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000008']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000008']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000008']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000008']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000008</div>
</div>
<div class="line">
  <div class="data" data-tag="inode" data-xpath="/top/data/entry[name = 'xx-00000009']/inode"> 12</div>
  <div class="text"> </div>
  <div class="data" data-tag="blocks" data-xpath="/top/data/entry[name = 'xx-00000009']/blocks">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="mode" data-xpath="/top/data/entry[name = 'xx-00000009']/mode">mode</div>
  <div class="text"> </div>
  <div class="data" data-tag="links" data-xpath="/top/data/entry[name = 'xx-00000009']/links">12</div>
  <div class="text"> </div>
  <div class="data" data-tag="user" data-xpath="/top/data/entry[name = 'xx-00000009']/user">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="group" data-xpath="/top/data/entry[name = 'xx-00000009']/group">phil</div>
  <div class="text">  </div>
  <div class="data" data-tag="flags" data-xpath="/top/data/entry[name = 'xx-00000009']/flags">123</div>
  <div class="text"> </div>
  <div class="data" data-tag="label" data-xpath="/top/data/entry[name = 'xx-00000009']/label">1234</div>
  <div class="text"> </div>
  <div class="data" data-tag="size" data-xpath="/top/data/entry[name = 'xx-00000009']/size">12345</div>
  <div class="text"> </div>
  <div class="data" data-tag="modify-time" data-xpath="/top/data/entry[name = 'xx-00000009']/modify-time"> 8 Feb 21:30</div>
  <div class="text"> </div>
  <div class="data" data-tag="name" data-xpath="/top/data/entry/name">xx-00000009</div>
</div>
<div class="line">
  <div class="label">hits</div>
  <div class="decoration">:</div>
  <div class="padding"> </div>
  <div class="data" data-tag="hits" data-xpath="/top/data/hits">72</div>
</div>
